engine/cld/offchain/offchain.go

```
@@ -2,83 +2,195 @@ package offchain
+func withTokenSource(tokenSource oauth2.TokenSource) LoadOffchainClientOpt {
+	return func(c *loadConfig) {
+		c.tokenSource = tokenSource
+	}
+}
+
 // LoadOffchainClient loads an offchain client for the specified domain and environment.
 func LoadOffchainClient(
 	ctx context.Context,
-	domain enginedomain.Domain,
-	env string,
-	config *cfgenv.Config,
-	lggr logger.Logger,
-	useRealBackends bool,
+	dom domain.Domain,
+	cfg cfgenv.JobDistributorConfig,
+	opts ...LoadOffchainClientOpt,
 ) (foffchain.Client, error) {
-	var jd foffchain.Client
+	loadCfg, err := defaultLoadConfig()
+	if err != nil {
+		return nil, err
+	}
+	for _, opt := range opts {
+		opt(loadCfg)
+	}
 
-	endpoints := config.Offchain.JobDistributor.Endpoints
-	auth := config.Offchain.JobDistributor.Auth
+	var (
+		lggr      = loadCfg.logger
+		endpoints = cfg.Endpoints
+		auth      = cfg.Auth
+	)
 
-	if domain.Key() == enginedomain.MustGetDomain("keystone").Key() && endpoints.WSRPC == "" {
+	// TODO: Remove this domain specific check
+	if dom.Key() == "keystone" && endpoints.WSRPC == "" {
 		lggr.Warn("Skipping JD initialization for Keystone, fallback to CLO data")
-	} else if endpoints.WSRPC == "" || endpoints.GRPC == "" {
-		lggr.Warn("Skipping JD initialization no JD config for WS or gRPC")
-	} else if endpoints.WSRPC != "" && endpoints.GRPC != "" {
-		lggr.Info("Initializing JD client")
-		var oauth oauth2.TokenSource
-		if config.Offchain.JobDistributor.Auth != nil {
-			source := jdoffchain.NewCognitoTokenSource(jdoffchain.CognitoAuth{
-				AppClientID:     auth.CognitoAppClientID,
-				AppClientSecret: auth.CognitoAppClientSecret,
-				Username:        auth.Username,
-				Password:        auth.Password,
-				AWSRegion:       auth.AWSRegion,
-			})
-			if err := source.Authenticate(ctx); err != nil {
+
+		return nil, nil //nolint:nilnil // We want to return nil if the JD is not initialized for now.
+	}
+
+	if endpoints.WSRPC == "" || endpoints.GRPC == "" {
+		return nil, ErrEndpointsRequired
```

Instead of skipping as previously implemented, we return a sentinel error to let the caller decide how they wish to proceed. There are valid cases where you may want to fail entirely instead of skipping the client load. (See the JD commands case)